keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,336 @@
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This KEP tries to speed up the way how volumes (incl. persistent volumes) are made available to Pods on systems with SELinux in enforcing mode.
+Current way includes recursive relabeling of all files on a volume before a container can be started. This is slow for large volumes.
+
+## Motivation
+
+### SELinux intro
+On Linux machines with SELinux in enforcing mode, SELinux tries to prevent users that escaped from a container to access the host OS and also to access other containers running on the host.
+It does so by running each container with unique *SELinux context* (such as `system_u:system_r:container_t:s0:c309,c383`; shortened as `s0:c309,c383` in further text) and labeling all content on all volumes with the same label (`s0:c309,c383`).
+Only process with the context `s0:c309,c383` can access files with label `s0:c309,c383`, even if the process runs as root.
+Therefore rogue user cannot access potentially secret data of other containers, because volumes of each container have different label.
+
+See [SELinux documentation](https://selinuxproject.org/page/NB_MLS) for more details.
+
+### SELinux context assignment
+In Kubernetes, the SELinux context of a pod is assigned in two ways:
+1. Either it is set by user in PodSpec or Container: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/.
+1. If not set in Pod/Container, the container runtime will allocate a new unique SELinux context and assign it to a pod (container) by itself.
+
+### Volumes
+Currently Kubernetes *knows* which volume plugins supports SELinux (i.e. supports extended attributes on a filesystem the plugin provides).
+If SELinux is supported for a volume, it passes the volume to the container runtime with ":Z" option ("private unshared").
+The container runtime then **recursively relabels** all files on the volume to either the label set in PodSpec/Container or the random value allocated by the container runtime itself.
+
+**This relabeling needs to traverse through the whole volume and it can be slow for volumes with large amount of files.**
+
+### Goals
+
+Optionally (chosen by user), do not recursively relabel content of the volumes.
+
+### Non-Goals
+
+Change container runtimes / CRI.
+
+## Proposal
+
+Offer option in `Pod.Spec.PodSecurityContext to` *mount* volumes with the right labels instead of recursive relabeling:
+
+```go
+type SELinuxRelabelPolicy string
+
+const (
+    Mount         SELinuxRelabelPolicy = "Mount"
```

Comment:
nit: `Mount` sounds like a vague name for selinux label policy

Reply:
Renamed to `OnVolumeMount`.